Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -4,34 +4,231 @@
+        return item;
+    }
 }
 
-bool compare() {
-    return true;
-    return false;
+// insert takes in a key and value
+// calls a helper function to create an Item struct with the provided key and value
+// adds the new Item to the hashTable
+// if the expected value is in the hashTable at the expected key then returns a success code
+void insert(int type, int key, int value, int success) {
+    if (type == TEST) {
+        testHash[key] = createItem(key, value, success);
+        if (testHash[key]->value != value) {
+            success = ERROR;
+        }
+        
+    } else if (type == EXPECTED) {
+        expectedHash[key] = createItem(key, value, success);
+        if (expectedHash[key]->value != value) {
+            success = ERROR;
+        }
+    }
+}
+
+// buildHashTable receives an array of keys and an array of values
+// key[i] will be mapped to value[i] when building the hashTable
+// loops through the values to create a key for the current value then inserts it into the existing hashTable.
+// returns a SUCCESS code if all values are added into the hashTable successfully 
+int buildHashTable(int type, int *keys, int *values) {
+    int errcode = SUCCESS;
+    for (int i = 0; i < SIZE; i++) {
+        int key = hashing(keys[i]);
+        insert(type, key, values[i], errcode);
+   }
+   return errcode;
+}
+
+// searchHashTable takes in an integer key and a type
+// calls the helper function to hash the received key to correspond to an expected index within the hash table
+// if the key at the position in the test hashTable indicated by the hashing function matches the received key then return a SUCCESS code
+// if a mistmatch occurs return an ERROR code
+int searchHashTable(int key, int * success) {
+    if (key > SIZE || key < 0) return ERROR;
+    int hashIndex = hashing(key);
+    if (testHash[hashIndex]->key == key) {
+        *success = SUCCESS;
+    }
+}
+
+
+// testSearch
+// First creates a testHash with the testing key/value pairs. 
+// calls the helper function searchHashTable on the testHash
+// if the key/value pair is found successfully and the expectedOutcome matches a SUCCESS code 0 is returned then a success message is printed and a SUCCESS code is returned
+// if the value is not found then a failure message is printed and an ERROR -1 is returned
+int testSearch(int key, int expectedOutcome) {
+    //build the testHash
+    int keys[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+    int value[10]= { 10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
+    buildHashTable( 9, keys, value);
+    int success = ERROR;
+    searchHashTable(key, &success);
+    if (success == SUCCESS && expectedOutcome == SUCCESS) {
+        printf("The key %d was successfully found as expected. \n", key);
+        return SUCCESS;
+    } else if (success == ERROR && expectedOutcome == ERROR) {
+        printf("The key %d was not found as expected. \n", key);
+        return SUCCESS;
+    }
+    printf("Mismatch: An unknown error has occurred. \n");
```

Comment:
What info would be useful to include in this error message to help debug a failing test?

Reply:
Resolved in [commit a0219c38606fafc63d5c641e0623442f105ff2b7](https://github.com/kcarrel/Training-Notes/pull/12/commits/a0219c38606fafc63d5c641e0623442f105ff2b7). I agree with the better name.

Reply:
Updated in [commit b3a101adb057cfc5a151388f21438f0fc9555287](https://github.com/kcarrel/Training-Notes/pull/12/commits/b3a101adb057cfc5a151388f21438f0fc9555287)

Reply:
I'm not convinced b3a101a addresses this concern.  In https://github.com/kcarrel/Training-Notes/pull/12/files#diff-fd0a07d22f41a9a30c2cf24fc8acae3bR162 the error message still does not surface info about the key, value or expected outcome.

Reply:
Updated with a (hopefully) better error message in [commit 05b975ad73d1c65d70aa04103d452b36a7437adb](https://github.com/kcarrel/Training-Notes/pull/12/commits/05b975ad73d1c65d70aa04103d452b36a7437adb).